src/db/gc/mod.rs

```diff
@@ -0,0 +1,435 @@
+//! considered correct when a valid snapshot can be exported using records available in the database
+//! after the run.
+//!
+//! ## Disk usage
+//! The expected disk usage is slightly greater than the size of live data for three reasons:
+//! 1. Unreachable data is not removed until it is at least 7.5 hours old (see `chain finality`).
+//! 2. The garbage collector is conservative and is expected to leave a small (less than 1%) amount
+//! of unreachable data behind.
+//! 3. The blockstore back-end may be fragmented, therefore not relinquishing the disk space back to
+//! the OS.
+//!
+//! ## Memory usage
+//! During the `mark` and up to the `sweep` stage, the algorithm requires `4 bytes` of memory for
+//! each database record. Additionally, the seen cache while traversing the reachable graph
+//! executing the `filter` stage requires at least `32 bytes` of memory for each reachable block.
+//! For a typical mainnet snapshot of about 100 GiB that adds up to roughly 2.5 GiB.
+//!
+//! ## Scheduling
+//! 1. GC is triggered automatically and there have to be at least `chain finality` epochs stored
+//! for the `mark` step.
+//! 2. The `filter` step is triggered after at least `chain finality` has passed since the `mark`
+//! step.
+//! 3. Then, the `sweep` step happens.
+//! 4. Finally, the algorithm waits for a configured amount of time to initiate the next run.
+//!
+//! ## Performance
+//! The time complexity of mark and sweep steps is `O(n)`. The filter step is currently utilizing a
+//! depth-first search algorithm, with `O(V+E)` complexity, where V is the number of vertices and E
+//! is the number of edges.
+
+use crate::blocks::Tipset;
+use crate::chain::ChainEpochDelta;
+
+use crate::db::{truncated_hash, GarbageCollectable};
+use crate::ipld::stream_graph;
+use crate::shim::clock::ChainEpoch;
+use ahash::{HashSet, HashSetExt};
+use futures::StreamExt;
+use fvm_ipld_blockstore::Blockstore;
+use std::mem;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time;
+use tracing::info;
+
+/// [`MarkAndSweep`] is a simple garbage collector implementation that traverses all the database
+/// keys writing them to a [`HashSet`], then filters out those that need to be kept and schedules
+/// the rest for removal.
+///
+/// Note: The GC does not know anything about the hybrid CAR-backed and ParityDB approach, only
+/// taking care of the latter.
+pub struct MarkAndSweep<DB> {
+    db: Arc<DB>,
+    get_heaviest_tipset: Box<dyn Fn() -> Arc<Tipset> + Send>,
+    marked: HashSet<u32>,
+    epoch_marked: ChainEpoch,
+    depth: ChainEpochDelta,
+    block_time: Duration,
+}
+
+impl<DB: Blockstore + GarbageCollectable> MarkAndSweep<DB> {
+    /// Creates a new mark-and-sweep garbage collector.
+    ///
+    /// # Arguments
+    ///
+    /// * `db` - A reference to the database instance.
+    /// * `get_heaviest_tipset` - A function that facilitates heaviest tipset retrieval.
+    /// * `depth` - The number of state-roots to retain.
```

Comment:
What should `depth` be initialized to? In the associated issue we're speaking of a number of `2 x chain_finality`, here we're taking max of `chain_finality` and `recent_state_roots`.

Reply:
It's not garbage collector's concern, the decision is made by the caller. But it's a good catch, and I will comment and fix it to be 2x chain_finality.